model-optimizer/extensions/middle/StridedSliceNormalizer.py

```diff
@@ -0,0 +1,243 @@
+from mo.ops.concat import Concat
+from mo.utils.error import Error
+from mo.ops.const import Const
+from mo.ops.op import PermuteAttrs
+
+
+class StridedSliceNormalizer(MiddleReplacementPattern):
+    """
+    StridedSlice is not normal if it cannot be permuted by ApplyPermutations. This normalizer
+    inserts blank colons ':' in slice expression so that it can be correctly permuted
+    from NHWC to NCHW layout. It changes masks and inserts blank begin, end and strides values.
+    In order to successfully handle StridedSlice in ShapeOf subgraphs
+    changes must be done by inserting nodes not just by overwriting constants.
+
+    StridedSlice is not normal in 2 cases:
+        1. rank of a slice expression is less than rank of input tensor
+        2. there is an ellipsis
+
+    1st case example
+    BEFORE:
+                  |
+                begin
+           value=[0, 0]
+                  |
+
+    AFTER:
+                  |
+                begin          Const
+           value=[0, 0]     value=[0, 0]
+                  \             /
+                   \           /
+                      Concat
+                 value=[0, 0, 0, 0]
+                        |
+
+    Input of a shape [16, 100, 100, 3] in NHWC layout, output = input[:, 0:50].
+    StridedSlice ill be extended to input[:, 0:50, :, :].
+    After permutation to NCHW output = input[:, :, 0:50, :].
+    Above is show only for begin input, for end and strides changes are analogous.
+
+    2nd case example
+    BEFORE:
+                  |
+                begin
+           value=[1, 50]
+                  |
+
+    AFTER:
+                  |
+                begin
+           value=[1, 1, 1]
+                  |
+             VariadicSplit
+           /              \
+          /                \
+         /       Const      \
+         \     val=[0, 0]   /
+          \       |        /
+           \      |       /
+               Concat
+           value=[1, 0, 0, 1, 1]
+                  |
+
+    Input of a shape [16, 10, 100, 100, 3] in NDHWC layout, output = input[1:4, ..., 1:51, 1:3],
+    output_shape = [3, 10, 100, 50, 2]. In order to perform correct layout permutation
+    ellipsis should be exended: input[1:4, ..., 1:51, 1:3] => input[1:4, :, :, 1:51, 1:3].
+    Afterc layour permutation input[1:4, 1:3, :, : 1:5].
+
+    In the places of colons blank zero begin, end and strides values
+    should be inserted. In order to do that we split begin, and concatenate with the blank zeros in the middle.
+    Above is show only for begin input, for end and strides changes are analogous.
+    """
+    enabled = True
+
+    def run_before(self):
+        from extensions.middle.LayoutChangeForConstantShapePaths import LayoutChangeForConstantShapePaths
+        return [LayoutChangeForConstantShapePaths]
+
+    def find_and_replace_pattern(self, graph: Graph):
+        for node in graph.get_op_nodes(op='StridedSlice'):
```

Comment:
should we use type?

Reply:
changed to `type` as it was used in other SS transformations before my changes